torch/distributed/algorithms/join.py

```diff
@@ -0,0 +1,210 @@
+import warnings
+from typing import List
+from abc import ABC, abstractmethod
+
+import torch
+import torch.distributed as dist
+
+class _JoinHook(ABC):
+    r"""
+    This defines a join hook, which provides two entry points in the join
+    context manager: a main hook, which is called repeatedly while there exists
+    a non-joined process, and a post-hook, which is called once all processes
+    have joined.
+
+    To implement a join hook for the generic join context manager, define a
+    class that inherits from :class:`_JoinHook`, override ``main_hook()`` and
+    ``post_hook()`` as appropriate, and override ``device()`` and
+    ``process_group()`` to provide the device and process group information,
+    respectively, which are needed for the join context manager implementation.
+    """
+    def main_hook(self):
+        r"""
+        This hook is called repeatedly while there exists a non-joined process
+        to shadow collective communications in the forward pass, backward pass,
+        and optimizer.
+        """
+        ...
+
+    def post_hook(self, is_last_joiner: bool):
+        r"""
+        This hook is called after all processes have joined. It is passed an
+        additional ``bool`` argument ``is_last_joiner``, which indicates if the
+        rank is one of the last to join.
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def device(self):
+        r"""
+        Returns the device from which to perform collective communications
+        needed for the join context manager implementation itself.
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def process_group(self):
+        r"""
+        Returns the process group for join-related collective communications.
+        """
+        ...
+
+class _Join():
+    r"""
+    This class defines the generic join context manager, which allows custom
+    hooks to be called after a process joins. These hooks should shadow the
+    collective communications of non-joined processes to prevent hanging and
+    erroring and to ensure algorithmic correctness. Refer to :class:`_JoinHook`
+    for details about the hook definition.
+
+    .. warning::
+        The context manager requires a ``dist.all_reduce(torch.ones(1))`` to be
```

Comment:
(this is for future discussion and improvements) Another option we can explore is to use the c10d Store (e.g., `CompareAndSet` API) instead of using AllReduce. Looks like all we need here is to know the number of non-joined processes, and AllReduce might be an overkill? cc @H-Huang

Reply:
@H-Huang and I discussed this offline. 

IIUC, the idea is to have a global variable counting the number of joined processes. When a process joins (i.e. enters the context manager), it increments the variable using `get()` and `compare_set()`, and joined processes can check the variable against the world size to know when all processes have joined using `get()`. 

However, I still have one issue with this idea, but I may just be misunderstanding. (@H-Huang, I thought it was resolved in our discussion, but on second thought, maybe not.) Using the all-reduce to get the number of non-joined processes at each iteration provides a synchronization point for all processes, which is actually useful. If the processes instead use the global variable and `compare_set()` method as described above, then I think processes may hang:

Suppose there are two processes, rank `0` and rank `1`. Suppose rank `0` has already joined, meaning the global `count = 1`, and consider the last iteration of rank `1`. The issue lies in if rank `0` checks `count` before rank `1` is able to increment `count` as it enters the context manager. Then, rank `0` will see `count = 1 < 2` and determine that not all processes have joined yet. It will proceed to its main hooks, scheduling collective communications that rank `1` will never participate in. Rank `1` will increment `count` to `2` and proceed to the post-hooks. 

Any solution I see to this requires a synchronization point before the joined processes start the collective communications in their main hooks. I imagine that any method for synchronizing (e.g. `dist.barrier()`) will not be much lighter weight than `dist.all_reduce()`.

Reply:
I this this is indeed a concern. `dist.barrier()` I didn't closely track today's barrier, I recall it used to be implemented as an empty allreduce, but a store-based barrier should work. A more efficient solution might be `compare_set(barrier=True)`, as it only needs one round-trip? Not yet sure if this is an reasonable idea.